File: 1196-114-1261-find-elements-in-a-contaminated-binary-tree/1196-114-1261-find-elements-in-a-contaminated-binary-tree.java

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class FindElements {
    private TreeNode root;
    private Set<Integer> st;

    public FindElements(TreeNode root) {
        this.root = root;
        this.st = new HashSet<>();
        recover(root, 0);
    }

    private void recover(TreeNode node, int val) {
        if (node == null) return;
        node.val = val;
        st.add(val);
        recover(node.left, 2 * val + 1);
        recover(node.right, 2 * val + 2);
    }

    public boolean find(int target) {
        return st.contains(target);
    }
}

File: 1196-114-1261-find-elements-in-a-contaminated-binary-tree/1196-114-1261-find-elements-in-a-contaminated-binary-tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class FindElements:
    def __init__(self, root: TreeNode):
        self.root = root
        self.st = set()
        self.recover(root, 0) 
    def recover(self, node, val):
        if not node:
            return
        node.val = val
        self.st.add(val)
        self.recover(node.left, 2 * val + 1)
        self.recover(node.right, 2 * val + 2)

    def find(self, target: int) -> bool:
        return target in self.st
    